### DISCUSSION TITLE
Notes on Twind runtime performance (observe/shim, just-in-time stylesheet creation)

### DISCUSSION THREAD

**Discussion by danielweck** [tw-in-js/twind#161] [General]
Discussion topic started in this "issue":

https://github.com/tw-in-js/twind/issues/157#issuecomment-808315583

(moved here to avoid polluting the main conversation thread)

**Fun test**: use Twind/shim to render https://tailwindcss.com (a really impressive landing page packed with cool interactives!)

___
___

**SPOILER ALERT: Twind shim/observe is slower than precompiled stylesheets, and on first impression it even seems much slower. However when configured appropriately, it is actually impressively fast! Details below.**
____
____

I "downloaded" a local copy of the website's landing page ("save as" from Chrome), and I ran Lighthouse's "performance" test from a localhost HTTP server:

# Desktop:

<img width="734" alt="Screenshot 2021-03-26 at 16 43 19" src="https://user-images.githubusercontent.com/625201/112665517-0e671b00-8e53-11eb-978b-3d290237b457.png">

<img width="734" alt="Screenshot 2021-03-26 at 16 53 30" src="https://user-images.githubusercontent.com/625201/112666397-0cea2280-8e54-11eb-87a5-d065aadaf422.png">

# Mobile: 

<img width="734" alt="Screenshot 2021-03-26 at 16 44 14" src="https://user-images.githubusercontent.com/625201/112665621-2b9be980-8e53-11eb-9831-77bbd961434d.png">

<img width="734" alt="Screenshot 2021-03-26 at 16 53 53" src="https://user-images.githubusercontent.com/625201/112666423-13789a00-8e54-11eb-8839-4034bef9e6d7.png">

____

Then, I removed the Tailwind CSS stylesheet, and I added:

`<script type="module" src="https://cdn.skypack.dev/twind/shim"></script>`

(Skypack CDN, but I ran a local test too, see further below)

**EDIT: this is a suboptimal method for loading Twind, due to the async / deferred nature of ECMAScript modules. Traditional script loading yields much better perceived and measured performance due to scheduling layout recalc in relation to the DOMContentLoaded event. Please make sure to read the entire thread below to see much improved performance figures :) **

# Desktop:

<img width="734" alt="Screenshot 2021-03-26 at 16 42 49" src="https://user-images.githubusercontent.com/625201/112665761-4f5f2f80-8e53-11eb-87cd-c1e32bc10b53.png">

<img width="734" alt="Screenshot 2021-03-26 at 16 58 40" src="https://user-images.githubusercontent.com/625201/112666975-a3b6df00-8e54-11eb-8606-2feae6d5b9cb.png">

# Mobile:

<img width="734" alt="Screenshot 2021-03-26 at 16 41 39" src="https://user-images.githubusercontent.com/625201/112665803-59812e00-8e53-11eb-85c3-b63cc564c2fd.png">

<img width="734" alt="Screenshot 2021-03-26 at 16 59 11" src="https://user-images.githubusercontent.com/625201/112667011-ad404700-8e54-11eb-857d-2cf2723658d1.png">

____

Finally, I replaced the Skypack CDN Twind/shim script link with locally-hosted copies:

# Desktop

<img width="734" alt="Screenshot 2021-03-26 at 17 09 16" src="https://user-images.githubusercontent.com/625201/112668391-5cc9e900-8e56-11eb-85b4-ee87589aea7d.png">

<img width="734" alt="Screenshot 2021-03-26 at 17 09 43" src="https://user-images.githubusercontent.com/625201/112668415-65222400-8e56-11eb-9db7-e52d6ea89dba.png">

# Mobile

<img width="734" alt="Screenshot 2021-03-26 at 17 10 20" src="https://user-images.githubusercontent.com/625201/112668441-6d7a5f00-8e56-11eb-9860-3cbf4ed334ca.png">

<img width="734" alt="Screenshot 2021-03-26 at 17 10 49" src="https://user-images.githubusercontent.com/625201/112668463-73704000-8e56-11eb-94b3-8f8519b61a4b.png">

Here's the performance flamegraph:

# Without Twind:

<img width="1009" alt="Screenshot 2021-03-26 at 17 18 46" src="https://user-images.githubusercontent.com/625201/112669283-6ef85700-8e57-11eb-9864-49639a00c308.png">

# With Twind:

(the activity graph on the left hand side is Twind's `observe` aka "shim" immediately after `DOMContentLoaded`, working its way through the DOM to build the stylesheet just-in-time)

<img width="1009" alt="Screenshot 2021-03-26 at 17 18 03" src="https://user-images.githubusercontent.com/625201/112669311-79b2ec00-8e57-11eb-87d5-8bf42df3ec39.png">

@sastan :

> These results are to be expected because the shim kicks in after DOMContentLoaded.
> For production sites we recommend critical CSS extraction.
> I'm currently experimenting with one or two ways to optimize this. One of them is:

```html
<script>
// Inline https://unpkg.com/twind/twind.umd.js
// Inline https://unpkg.com/twind/observe/observe.umd.js

twindObserve.observe(document.documentElement)
</script>
```

> Could you try that one?

Sure. My current test is with:

```html
  <!-- script type="module" src="https://cdn.skypack.dev/twind/shim"></script -->
<!-- LOCALHOST COPY: -->
  <script type="module" src="./shim.js"></script>
```

___
___
___

# Much better :)

___
___
___

# Desktop

<img width="1009" alt="Screenshot 2021-03-26 at 17 27 51" src="https://user-images.githubusercontent.com/625201/112670351-b7fcdb00-8e58-11eb-82ff-546820807842.png">

<img width="1009" alt="Screenshot 2021-03-26 at 17 28 12" src="https://user-images.githubusercontent.com/625201/112670368-bcc18f00-8e58-11eb-8e9f-f55864593246.png">

# Mobile

<img width="1009" alt="Screenshot 2021-03-26 at 17 30 03" src="https://user-images.githubusercontent.com/625201/112670567-fbefe000-8e58-11eb-8bff-535499a6f6b7.png">

<img width="1009" alt="Screenshot 2021-03-26 at 17 30 26" src="https://user-images.githubusercontent.com/625201/112670583-00b49400-8e59-11eb-9254-f9e944221540.png">

Interesting flamegraph now:

<img width="1009" alt="Screenshot 2021-03-26 at 17 32 48" src="https://user-images.githubusercontent.com/625201/112671403-052d7c80-8e5a-11eb-8a5e-c48ea024f4bb.png">

**Comment by danielweck** [comment]
@sastan you can see in the performance flamegraph (immediately above) the `DCL` event (DOMContentLoaded) in the blue box on the right hand side (the `hydrate` phase is immediately after), and some fragmented activity before `DCL`. I interpret this as the HTML markup being parsed and `observe` triggering incrementally, as opposed to a single pass *after* `DCL`? (see the other flamegraph further above)

**Reply by sastan** [reply]
Really interesting. The style re-calc is killing us. This is all very good to know and we can start optimizing the observe on the initial load.

**Reply by danielweck** [reply]
I replicated your UMD method with ESModule:

```html
  <!-- script type="module" src="https://cdn.skypack.dev/twind/shim"></script -->
  <!-- script type="module" src="./shim.js"></script -->
  <script type="module">
    import { createObserver } from 'observe.js';
    const twindObserver = createObserver();
    twindObserver.observe(document.documentElement);
  </script>
```

**Reply by sastan** [reply]
There is a slight difference. `<script type="module">` are `defer` by default. The inline script is executed right away. I do not know what impact the difference has.

**Reply by danielweck** [reply]
Correction: the performance metrics reported by Lighthouse are NOT as good with the `type="module"` method, and the flamegraph does look quite different.

**Reply by danielweck** [reply]
<img width="1009" alt="Screenshot 2021-03-26 at 18 17 02" src="https://user-images.githubusercontent.com/625201/112675802-820f2500-8e5f-11eb-91fd-3098279ea76d.png">

**Reply by sastan** [reply]
The `<script>/* inline JS */</script>` method is better considering the flamegraph?

**Reply by ggoodman** [reply]
Could the shim observer's handler be batched into a deferred `requestAnimationFrame`? Is it already?

**Reply by sastan** [reply]
We need to generate all styles when they get applied to the DOM. Otherwise, the UI may be broken.

**Reply by sastan** [reply]
But we could try to detect if there is twind CSS in the page from (critical CSS extraction on the server) and prevent the initial run you see in the graph and go straight to _just_ observing.

**Reply by danielweck** [reply]
Regardless of what I see in the flamegraph, the FOUC Flash Of Unstyled Content in the webpage is **really** bad (feels like 500ms - 1s) with the ESM method, whereas it is near invisible with the UMD method.

**Reply by danielweck** [reply]
Yeah, and my visual assessment matches the Lighthouse score too. I retract my previous statement:
> The performance metrics reported by Lighthouse are equally good with the type="module" method

(I made a mistake in my measurements)

**Reply by danielweck** [reply]
I should also explain that I tried:

```html
  <script type="module">
    import { createObserver } from 'observe.js';
    const twindObserver = createObserver();
    twindObserver.observe(document.documentElement);
  </script>
```

and (`async`):

```html
<script type="module" src="tw.js" async> </script>
```

and (`defer`):

```html
<script type="module" src="tw.js" defer> </script>
```

...with `tw.js`:
```js
    import { createObserver } from 'observe.js';
    const twindObserver = createObserver();
    twindObserver.observe(document.documentElement);
```

=> same results (`async` is more likely to kick in after `DOMContentLoaded`, whereas `defer` in principle occurs before the `DCL` event).

**Reply by ggoodman** [reply]
>But we could try to detect if there is twind CSS in the page from (critical CSS extraction on the server) and prevent the initial run you see in the graph and go straight to just observing.

This sounds like something that could dramatically reduce the amount of _work_ needed in an SSR -> CSR transition. Love it!

**Reply by danielweck** [reply]
This is equally performant, no need to inline the scripts:

```html
  <script type="text/javascript" crossorigin="anonymous" src="https://unpkg.com/twind@0.16.9/twind.umd.js"> </script>

  <script type="text/javascript" crossorigin="anonymous" src="https://unpkg.com/twind@0.16.9/observe/observe.umd.js"> </script>
  <script  type="text/javascript">
    twindObserve.observe(document.documentElement)
    </script>
```

**Reply by danielweck** [reply]
As you can see, when fetching the scripts from the `unpkg` CDN, the pre-hydration activity is less fragmented:


<img width="1009" alt="Screenshot 2021-03-26 at 18 51 30" src="https://user-images.githubusercontent.com/625201/112679797-7f62fe80-8e64-11eb-83a0-a8114b751ac6.png">

**Comment by danielweck** [comment]
@sastan I added the following paragraph in the introduction post, to avoid misleading readers:

> SPOILER ALERT: Twind shim/observe is slower than precompiled stylesheets, and on first impression it even seems much slower. However when configured appropriately, it is actually impressively fast! Details below.

**Comment by danielweck** [comment]
I also added the following clarification in the thread:

**EDIT: this is a suboptimal method for loading Twind, due to the async / deferred nature of ECMAScript modules. Traditional script loading yields much better perceived and measured performance due to scheduling layout recalc in relation to the DOMContentLoaded event. Please make sure to read the entire thread below to see much improved performance figures :) **

**Reply by sastan** [reply]
I implemented the umd+observe method on twind.dev

Before: https://lightest.app/c/HpfJQxGTN2

![Bildschirmfoto 2021-03-29 um 09 04 02](https://user-images.githubusercontent.com/514405/112798798-b821f400-906d-11eb-8d1d-f37572520f19.png)

After: https://lightest.app/c/WrH41q_VYB

![Bildschirmfoto 2021-03-29 um 09 03 50](https://user-images.githubusercontent.com/514405/112798776-af312280-906d-11eb-8989-3b5dc9633cd1.png)

**Reply by danielweck** [reply]
...looks like Twind's doc pages (TypeDoc-generated) could be much leaner :)

https://estimator.dev/#twind.dev/docs/

See:

https://github.com/tw-in-js/twind/tree/main/docs

=>

https://github.com/sastan/typedoc-plugin-pages
+
https://github.com/TypeStrong/TypeDoc

**Reply by sastan** [reply]
Yes. working on in the docs #124 PR

**Reply by danielweck** [reply]
> Yes. working on in the docs #124 PR

Oooh, very nice! 👍 Thanks for the "heads up".
